pkg/contract/chain.go

```diff
@@ -5,50 +5,336 @@ package contract
+	"github.com/ava-labs/avalanche-cli/pkg/utils"
 	"github.com/ava-labs/avalanchego/ids"
 	"github.com/spf13/cobra"
 )
 
-type ChainFlags struct {
-	SubnetName string
-	CChain     bool
+type ChainSpec struct {
+	BlockchainName string
+	CChain         bool
+	BlockchainID   string
 }
 
-func AddChainFlagsToCmd(
+func MutuallyExclusiveChainSpecFields(chainSpec ChainSpec) bool {
+	return cmdflags.EnsureMutuallyExclusive([]bool{
+		chainSpec.BlockchainName != "",
+		chainSpec.BlockchainID != "",
+		chainSpec.CChain,
+	})
+}
+
+func DefinedChainSpec(chainSpec ChainSpec) bool {
+	return chainSpec.BlockchainName != "" || chainSpec.BlockchainID != "" || chainSpec.CChain
+}
+
+func AddChainSpecToCmd(
 	cmd *cobra.Command,
-	chainFlags *ChainFlags,
+	chainSpec *ChainSpec,
 	goal string,
-	subnetFlagName string,
+	blockchainFlagName string,
 	cChainFlagName string,
+	blockchainIDFlagName string,
+	addBlockchainIDFlag bool,
 ) {
-	if subnetFlagName == "" {
+	subnetFlagName := "" // backwards compat
+	if blockchainFlagName == "" {
+		blockchainFlagName = "blockchain"
 		subnetFlagName = "subnet"
 	}
 	if cChainFlagName == "" {
 		cChainFlagName = "c-chain"
 	}
-	cmd.Flags().StringVar(&chainFlags.SubnetName, subnetFlagName, "", fmt.Sprintf("%s into the given CLI subnet", goal))
-	cmd.Flags().BoolVar(&chainFlags.CChain, cChainFlagName, false, fmt.Sprintf("%s into C-Chain", goal))
+	if blockchainIDFlagName == "" {
+		blockchainIDFlagName = "blockchain-id"
+	}
+	if subnetFlagName != "" {
+		cmd.Flags().StringVar(&chainSpec.BlockchainName, subnetFlagName, "", fmt.Sprintf("%s into the given CLI blockchain", goal))
+	}
+	cmd.Flags().StringVar(&chainSpec.BlockchainName, blockchainFlagName, "", fmt.Sprintf("%s into the given CLI blockchain", goal))
+	cmd.Flags().BoolVar(&chainSpec.CChain, cChainFlagName, false, fmt.Sprintf("%s into C-Chain", goal))
+	if addBlockchainIDFlag {
+		cmd.Flags().StringVar(&chainSpec.BlockchainID, blockchainIDFlagName, "", fmt.Sprintf("%s into the given blockchain ID/Alias", goal))
+	}
+}
+
+func GetBlockchainEndpoints(
+	app *application.Avalanche,
+	network models.Network,
+	chainSpec ChainSpec,
+	promptForRPCEndpoint bool,
+	promptForWSEndpoint bool,
+) (string, string, error) {
+	var (
+		rpcEndpoint string
+		wsEndpoint  string
+	)
+	switch {
+	case chainSpec.CChain:
+		rpcEndpoint = network.CChainEndpoint()
+		wsEndpoint = network.CChainWSEndpoint()
+	case network.Kind == models.Local || network.Kind == models.Devnet:
+		blockchainID, err := GetBlockchainID(app, network, chainSpec)
+		if err != nil {
+			return "", "", err
+		}
+		rpcEndpoint = network.BlockchainEndpoint(blockchainID.String())
+		wsEndpoint = network.BlockchainWSEndpoint(blockchainID.String())
+	case chainSpec.BlockchainName != "":
+		sc, err := app.LoadSidecar(chainSpec.BlockchainName)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to load sidecar: %w", err)
+		}
+		if sc.Networks[network.Name()].BlockchainID == ids.Empty {
```

Comment:
we should check for `sc.Networks[network.Name()]` key in map existance to avoid crash

Reply:
https://github.com/ava-labs/avalanche-cli/pull/2107#discussion_r1749295060